src/sage/graphs/matching_covered_graph.py

```diff
@@ -2003,6 +2004,361 @@ def has_perfect_matching(G, algorithm='Edmonds', solver=None, verbose=0,
+            return (True, None) if coNP_certificate else True
+
+        color = {u: 0 if u in self.bipartite_sets()[0] else 1 for u in self}
+        matching = set(self.get_matching())
+        matching_neighbor = {x: y for u, v, *_ in self.get_matching() for x, y in [(u, v), (v, u)]}
+
+        for e in list(self.get_matching())[:]:
+            u, v, *_ = e
+
+            # Let G denote the undirected graph self, and
+            # let the graph H(e) := G — u — v
+            H = Graph(self, multiedges=False)
+            H.delete_vertices([u, v])
+
+            if not H.is_matching_covered(list(matching - set([e]))):
+                if not coNP_certificate:
+                    return False
+
+                # Construct the digraph D(e)(A ∪ B, F) defined as follows:
+                from sage.graphs.digraph import DiGraph
+                D = DiGraph()
+
+                # For each edge (a, b) in E(H(e)) ∩ M with a in A, b —> a in D(e).
+                # For each edge (a, b) in E(H(e)) with a in A, a —> b in D(e).
+                for a, b, *_ in H.edge_iterator():
+
+                    if color[a]:
+                        a, b = b, a
+
+                    D.add_edge((a, b))
+                    if matching_neighbor[a] == b:
+                        D.add_edge((b, a))
+
+                D.show()
+
+                # H(e) is matching covered iff D(e) is strongly connected.
+                # Check if D(e) is strongly connected using Kosaraju's algorithm
+                def dfs(J, v, visited, orientation):
+                    stack = [v]  # a stack of vertices
+
+                    while stack:
+                        v = stack.pop()
+
+                        if v not in visited:
+                            visited[v] = True
+
+                        if orientation == 'in':
+                            for u in J.neighbors_out(v):
+                                if u not in visited:
+                                    stack.append(u)
+
+                        elif orientation == 'out':
+                            for u in J.neighbors_in(v):
+                                if u not in visited:
+                                    stack.append(u)
+                        else:
+                            raise ValueError('Unknown orientation')
+
+                root = next(D.vertex_iterator())
+
+                visited_in = {}
+                dfs(D, root, visited_in, 'in')
+
+                visited_out = {}
+                dfs(D, root, visited_out, 'out')
+
+                # Note that by definition of D(e), it follows that C ⊆ E(H(e)) — M.
+                # Thus, C is a cut of H(e), which has a shore X such that every edge of C is
+                # incident with a vertex in X ∩ B.
+
+                # Moreover, M — e is a perfect matching of H(e), and thus, |X ∩ A| = |X ∩ B|
+                # Consequently, Y := X + v is a shore of a nontrivial tight cut T of G
+
+                if not all(visited_in.values()):
```

Comment:
since you never assign `False`, all values are `True` and so `not all(visited_in.values())` is always `False`.

Reply:
Ideally, it should have been, `X` corresponds to `visited_in` if not all the vertices in `D` are in `visited_in` otherwise it corresponds to `visited_out`.

Seems like I need to add more doctests covering this line.